Indicators/HeikinAshi.cs

```diff
@@ -0,0 +1,117 @@
+    /// </summary>
+    public class HeikinAshi : TradeBarIndicator
+    {
+        /// <summary>
+        /// Gets the Heikin-Ashi Open
+        /// </summary>
+        public IndicatorBase<IndicatorDataPoint> Open { get; private set; }
+
+        /// <summary>
+        /// Gets the Heikin-Ashi High
+        /// </summary>
+        public IndicatorBase<IndicatorDataPoint> High { get; private set; }
+
+        /// <summary>
+        /// Gets the Heikin-Ashi Low
+        /// </summary>
+        public IndicatorBase<IndicatorDataPoint> Low { get; private set; }
+
+        /// <summary>
+        /// Gets the Heikin-Ashi Close
+        /// </summary>
+        public IndicatorBase<IndicatorDataPoint> Close { get; private set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="HeikinAshi"/> class using the specified name.
+        /// </summary> 
+        /// <param name="name">The name of this indicator</param>
+        public HeikinAshi(string name)
+            : base(name)
+        {
+            Open = new Identity(name + "_Open");
+            High = new Identity(name + "_High");
+            Low = new Identity(name + "_Low");
+            Close = new Identity(name + "_Close");
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="HeikinAshi"/> class.
+        /// </summary> 
+        public HeikinAshi()
+            : this("HA")
+        {
+        }
+
+        /// <summary>
+        /// Gets a flag indicating when this indicator is ready and fully initialized
+        /// </summary>
+        public override bool IsReady
+        {
+            get { return Samples > 1; }
+        }
+
+        /// <summary>
+        /// Computes the next value of this indicator from the given state
+        /// </summary>
+        /// <param name="input">The input given to the indicator</param>
+        /// <returns> A new value for this indicator </returns>
+        protected override decimal ComputeNextValue(TradeBar input)
+        {
+            if (!IsReady)
+            {
+                Open.Update(new IndicatorDataPoint(input.Time, (input.Open + input.Close) / 2));
+                Close.Update(new IndicatorDataPoint(input.Time, (input.Open + input.High + input.Low + input.Close) / 4));
+                High.Update(new IndicatorDataPoint(input.Time, input.High));
+                Low.Update(new IndicatorDataPoint(input.Time, input.Low));
+            }
+            else
+            {
+                Open.Update(new IndicatorDataPoint(input.Time, (Open + Close) / 2));
+                Close.Update(new IndicatorDataPoint(input.Time, (input.Open + input.High + input.Low + input.Close) / 4));
+                High.Update(new IndicatorDataPoint(input.Time, Math.Max(input.High, Math.Max(Open, Close))));
+                Low.Update(new IndicatorDataPoint(input.Time, Math.Min(input.Low, Math.Min(Open, Close))));
+            }
+
+            return Close;
```

Comment:
To build each output bar it needs the current bar and the previous one.
In the link below, only for the first bar they use a different formula.

More info: http://stockcharts.com/school/doku.php?id=chart_school:chart_analysis:heikin_ashi